TFUtil.py

```diff
@@ -1249,6 +1249,54 @@ def copy_template(self, name=None, dtype=None):
       kwargs["dtype"] = dtype
     return Data(**kwargs)
 
+  def copy_template_excluding_axis(self, exclude_axis, name=None):
+    """
+    :param str|None name: if set, this will be the new name.
+    :param int exclude_axis: axis to be removed.
+    :return: copy of myself excluding remove_axis_num axis, without placeholder.
+    :rtype: Data
+    """
+    kwargs = self.get_kwargs()
+    if exclude_axis < 0:
+      exclude_axis += self.batch_ndim
+      assert exclude_axis >= 0
+    assert 0 <= exclude_axis < self.batch_ndim
+    exclude_batch_dim_axis = exclude_axis == self.batch_dim_axis
+    if not exclude_batch_dim_axis:
+      axis_to_exclude_wo_b = self.get_batch_axis_excluding_batch(exclude_axis)
+    if exclude_axis == self.feature_dim_axis:
+      kwargs["dim"] = None
+
+    other_special_axes = self.get_special_axes_dict(
+      counted_with_batch_dim=True, only_available=True, include_batch_dim_axis=True)
+    for axis_name, axis in other_special_axes.items():
+      if axis == exclude_axis:
+        kwargs[axis_name] = None
```

Comment:
Better: `del kwargs[axis_name]`.
(E.g. when you remove the feature dim axis, or time dim axis, maybe some other spatial axis would automatically become the new time dim axis. `Data.__init__` should handle this.)

Reply:
with `del kwargs[axis_name]`, the test fails with:
`axis time_dim_axis (0) must be different from batch_dim_axis (0)`
when the batch_dim_axis is removed.
Am I doing something wrong in my test?
(also maybe I should split the current test into separate tests testing the removal of each special axis?).

Reply:
What is the full error exactly? And the relevant stack?

Reply:
```
Executing: test_Data_copy_template_excluding_axis
EXCEPTION
Traceback (most recent call last):
  File "tests/test_TFUtil.py", line 3242, in <module>
    line: globals()[arg]()  # assume function and execute
    locals:
      globals = <builtin> <built-in function globals>
      arg = <local> 'test_Data_copy_template_excluding_axis', len = 38
  File "tests/test_TFUtil.py", line 310, in test_Data_copy_template_excluding_axis
    line: data_wo_batch = data.copy_template_excluding_axis(data.batch_dim_axis)
    locals:
      data_wo_batch = <not found>
      data = <local> Data(name='data', shape=(None, 8), feature_dim_axis=2, batch_shape_meta=[B,T|?,F|8])
      data.copy_template_excluding_axis = <local> <bound method Data.copy_template_excluding_axis of Data(name='data', shape=(None, 8), feature_dim_axis=2, batch_shape_meta=[B,T|?,F|8])>
      data.batch_dim_axis = <local> 0
  File "/u/irie/src/returnn_ki/TFUtil.py", line 1293, in copy_template_excluding_axis
    line: return Data(**kwargs)
    locals:
      Data = <global> <class 'TFUtil.Data'>
      kwargs = <local> {'name': 'data', 'shape': [None, 8], 'dtype': 'float32', 'sparse': False, 'dim': 8, 'time_dim_axis': 0, 'feature_dim_axis': 1}, len = 7
  File "/u/irie/src/returnn_ki/TFUtil.py", line 477, in __init__
    line: self.sanity_check()
    locals:
      self = <local> Data(name='data', shape=(None, 8), feature_dim_axis=1, batch_shape_meta=[B|T,F|?,8])
      self.sanity_check = <local> <bound method Data.sanity_check of Data(name='data', shape=(None, 8), feature_dim_axis=1, batch_shape_meta=[B|T,F|?,8])>
  File "/u/irie/src/returnn_ki/TFUtil.py", line 499, in sanity_check
    line: assert axis != self.batch_dim_axis, "%s: axis %s (%i) must be different from batch_dim_axis (%i)" % (
            self, axis_name, axis, self.batch_dim_axis)
    locals:
      axis = <local> 0
      self = <local> Data(name='data', shape=(None, 8), feature_dim_axis=1, batch_shape_meta=[B|T,F|?,8])
      self.batch_dim_axis = <local> 0
      axis_name = <local> 'time_dim_axis', len = 13
AssertionError: Data(name='data', shape=(None, 8), feature_dim_axis=1, batch_shape_meta=[B|T,F|?,8]): axis time_dim_axis (0) must be different from batch_dim_axis (0)
```

Reply:
Ah I see. The default handling for `batch_dim_axis` is different. So it means you need to treat it different.
You need this addition code (after this loop):
```
if exclude_axis == self.batch_dim_axis:
  kwargs["batch_dim_axis"] = None
```